### PR TITLE
OG-125 configure registration block rate in server

### PR DESCRIPTION
**Pull request by shahafn**

### CODE REVIEW

File: src/relayserver/runServer.ts [hunk 1]
```diff
@@ -68,6 +69,8 @@ const ethereumNodeUrl: string = argv.ethereumNodeUrl ?? config.ethereumNodeUrl ?
 const workdir: string = argv.workdir ?? config.workdir ?? error('missing --workdir')
 const devMode: boolean = argv.devMode ?? config.devMode ?? error('missing --devMode')
 const debug: boolean = argv.debug ?? config.debug ?? error('missing --debug')
+const registrationBlockRate: string = argv.registrationBlockRate ?? config.registrationBlockRate?.toString()
+console.log('wtf is registrationBlockRate?', registrationBlockRate)
```
Comment (drortirosh) [inline src/relayserver/runServer.ts:73]: please be respectful with "wtf". if you do need it, put it in "debug"

Comment (drortirosh) [reply src/relayserver/runServer.ts:73]: or better: when the server starts, dump its complete active configuration.

File: test/RelayServer.test.ts [hunk 2]
```diff
@@ -1109,43 +1111,72 @@ contract('RelayServer', function (accounts) {
...
-        assertRelayAdded(receipts, newServer, false)
-        // @ts-ignore
-        newServer.pctRelayFee++
-        receipts = await newServer._registerIfNeeded()
-        assertRelayAdded(receipts, newServer, false)
-        // @ts-ignore
-        newServer.url = 'fakeUrl'
-        receipts = await newServer._registerIfNeeded()
-        assertRelayAdded(receipts, newServer, false)
+      describe('with re-registration', function () {
```
Comment (drortirosh) [inline test/RelayServer.test.ts:1160]: I'm trying to understand why you re-run all tests again, where the only test that can be affected is the added one.
why not just modify its configuration to see that (say) after 900 blocks i doens't re-register, and after 1100 it does ?

File: src/relayserver/runServer.ts (hunk 1 above)
Comment (shahafn) [reply src/relayserver/runServer.ts:73]: Or better, have a git hook not to commit it

File: src/relayserver/RelayServer.ts [hunk 3]
```diff
@@ -86,9 +88,12 @@ export interface RelayServerParams {
   readonly baseRelayFee: number | undefined
   readonly pctRelayFee: number | undefined
   readonly gasPriceFactor: number
+  readonly registrationBlockRate?: number
   readonly url?: string
   readonly workerMinBalance: number | undefined // = defaultWorkerMinBalance,
   readonly workerTargetBalance: number | undefined // = defaultWorkerTargetBalance,
+  readonly managerTargetBalance: number | undefined // = defaultManagerTargetBalance,
+  readonly minHubWithdrawalBalance: number | undefined // = defaultManagerTargetBalance,
```
Comment (forshtat) [inline src/relayserver/RelayServer.ts:96]: You use  '?:' and '| undefined' interchangeably. I think these are equivalent and it's better to stick to one syntax in general...

File: src/relayserver/RelayServer.ts [hunk 4]
```diff
@@ -471,19 +488,19 @@ export class RelayServer extends EventEmitter {
       throw new StateError('Could not get gasPrice from node')
     }
     const balance = await this.getManagerBalance()
-    if (balance.lt(toBN(minimumRelayBalance))) {
+    if (balance.lt(toBN(defaultMinManagerBalance))) {
       throw new StateError(
         // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
-        `Server's balance too low ( ${balance}, required ${minimumRelayBalance}). Waiting for funding...`)
+        `Server's balance too low ( ${balance}, required ${defaultMinManagerBalance}). Waiting for funding...`)
```
Comment (forshtat) [inline src/relayserver/RelayServer.ts:494]: Why did you change it to 'default'? Doesn't it break the sole idea of having it be configurable?

File: src/relayserver/RelayServer.ts [hunk 5]
```diff
@@ -644,17 +664,18 @@ export class RelayServer extends EventEmitter {
     return receipts
   }
 
-  async _isRegistered (): Promise<boolean> {
+  async _getRegistrationBlock (): Promise<number> {
```
Comment (forshtat) [inline src/relayserver/RelayServer.ts:667]: I was under impression the idea was for the server to send a re-register transaction in case no other transaction was made for a 'registrationBlockRate', so that the server does not become invisible. 
Here it will re-register even if active because it only checks the lates register event.
Is this on purpose?

File: src/relayserver/runServer.ts [hunk 6]
```diff
@@ -68,6 +69,7 @@ const ethereumNodeUrl: string = argv.ethereumNodeUrl ?? config.ethereumNodeUrl ?
 const workdir: string = argv.workdir ?? config.workdir ?? error('missing --workdir')
 const devMode: boolean = argv.devMode ?? config.devMode ?? error('missing --devMode')
 const debug: boolean = argv.debug ?? config.debug ?? error('missing --debug')
+const registrationBlockRate: string = argv.registrationBlockRate ?? config.registrationBlockRate?.toString()
```
Comment (forshtat) [inline src/relayserver/runServer.ts:72]: `null?.toString()` will return 'undefined' and this should break the types system here...

File: src/relayserver/RelayServer.ts (hunk 5 above)
Comment (shahafn) [reply src/relayserver/RelayServer.ts:669]: Fixed

File: src/relayserver/runServer.ts (hunk 6 above)
Comment (shahafn) [reply src/relayserver/runServer.ts:73]: Wouldn't that crash?

File: src/relayserver/RelayServer.ts (hunk 4 above)
Comment (shahafn) [reply src/relayserver/RelayServer.ts:495]: Actually I didn't change it, it was never configurable and I forgot to add it to configuration. Fixed.

File: src/relayserver/RelayServer.ts [hunk 7]
```diff
@@ -949,4 +986,8 @@ export class RelayServer extends EventEmitter {
     }
     return 1000
   }
+
+  address2topic (address: string): string {
```
Comment (forshtat) [inline src/relayserver/RelayServer.ts:990]: We already have 'event2topic' in the Utils.ts
I believe it will be beneficial to keep these utility functions in the same utility file instead of inside the server

File: src/relayserver/runServer.ts (hunk 6 above)
Comment (forshtat) [reply src/relayserver/runServer.ts:73]: And it will!..

File: test/RelayServer.test.ts (hunk 2 above)
Comment (forshtat) [reply test/RelayServer.test.ts:1189]: I don't understand the question but I don't see an answer either 😀

File: src/relayserver/RelayServer.ts (hunk 3 above)
Comment (forshtat) [reply src/relayserver/RelayServer.ts]: Nothing changed here

Comment (shahafn) [reply src/relayserver/RelayServer.ts]: This is true in general in the code, since it's not only me writing it. Consistency is important but should be fixed in another PR imo.

File: src/relayserver/RelayServer.ts (hunk 7 above)
Comment (shahafn) [reply src/relayserver/RelayServer.ts]: Fixed

File: test/RelayServer.test.ts (hunk 2 above)
Comment (shahafn) [reply test/RelayServer.test.ts:1185]: Fixed

**Comment by forshtat** [comment]
Unresolved issues migrated to OG-174
